src/Algebra/Graph/AdjacencyMap/Algorithm.hs

```diff
@@ -41,97 +41,73 @@ import Algebra.Graph.AdjacencyMap
-    True -> (:) <$> unfoldTreeM_BF walk v <*> explore vs
-    False -> explore vs
-  explore [] = return []
+bfsForest :: Ord a => [a] -> AdjacencyMap a -> Forest a
+bfsForest vs g = evalState (explore [ v | v <- vs, hasVertex v g ]) Set.empty where
+  explore = unfoldForestM_BF walk <=< filterM discovered
   walk v = (v,) <$> adjacentM v
   adjacentM v = filterM discovered $ Set.toList (postSet v g)
   discovered v = do new <- gets (not . Set.member v)
                     when new $ modify' (Set.insert v)
                     return new
 
--- | Like 'bfsForestFrom' with the resulting forest converted to a
---   level structure.  Flattening the result via @'concat' . 'bfs' vs@
---   gives an enumeration of vertices reachable from @vs@ in breadth
---   first order. Adjacent vertices are explored in increasing order
---   with respect to their 'Ord' instance.
+-- | This is 'bfsForest' with the resulting forest converted to a
+--   level structure. Adjacent vertices are explored in increasing
+--   order with respect to their 'Ord' instance. Flattening the result
+--   via @'concat' . 'bfs' vs@ gives an enumeration of vertices
+--   reachable from @vs@ in breadth first order.
 --
---   Let /L/ be the number of seed vertices. Complexity: /O((L+m)*log n)/
---   time and /O(n)/ space.
+--   Let /L/ be the number of seed vertices. Complexity:
+--   /O((L+m)*log n)/ time and /O(n)/ space.
 -- 
 -- @
 -- bfs vs 'empty'                                         == []
 -- bfs [] g                                             == []
--- bfs [1] ('edge' 1 1)                                   == [[1]]
--- bfs [1] ('edge' 1 2)                                   == [[1],[2]]
--- bfs [2] ('edge' 1 2)                                   == [[2]]
--- bfs [3] ('edge' 1 2)                                   == []
--- bfs [1,2] ('edge' 1 2)                                 == [[1],[2]]
--- bfs [2,1] ('edge' 1 2)                                 == [[2],[1]]
--- bfs [1,2] ( (1*2) + (3*4) + (5*6) )                  == [[1],[2]]
--- bfs [1,3] ( (1*2) + (3*4) + (5*6) )                  == [[1],[2],[3],[4]]
+-- bfs [1]   ('edge' 1 1)                                 == [[1]]
+-- bfs [1]   ('edge' 1 2)                                 == [[1],[2]]
+-- bfs [2]   ('edge' 1 2)                                 == [[2]]
+-- bfs [1,2] ('edge' 1 2)                                 == [[1,2]]
+-- bfs [2,1] ('edge' 1 2)                                 == [[2,1]]
+-- bfs [3]   ('edge' 1 2)                                 == []
+-- bfs [1,2] ( (1*2) + (3*4) + (5*6) )                  == [[1,2]]
+-- bfs [1,3] ( (1*2) + (3*4) + (5*6) )                  == [[1,3],[2,4]]
 -- bfs [3] (3 * (1 + 4) * (1 + 5))                      == [[3],[1,4,5]]
 -- bfs [2] ('circuit' [1..5] + 'circuit' [5,4..1])          == [[2],[1,3],[5,4]]
 -- 'concat' (bfs [3] $ 'circuit' [1..5] + 'circuit' [5,4..1]) == [3,2,4,1,5]
+-- bfs vs == 'map' 'concat' . 'List.transpose' . 'map' 'levels' . 'bfsForest' vs
```

Comment:
I didn't find the corresponding test, but we should have one: in future, we might change the implementation but we'd still need to make sure this property holds.

Reply:
Ah yes, that makes a lot of sense! I was initially thinking it was unnecessary to write a test that mirrored exactly the implementation